sdk/go/pulumi/types.go

```diff
@@ -547,8 +547,6 @@ func UnsafeUnknownOutput(deps []Resource) Output {
 func ToSecretWithContext(ctx context.Context, input interface{}) Output {
 	x := true
 	o := toOutputWithContext(ctx, nil, input, &x)
-	// set immediate secretness ahead of resolution/fufillment
-	o.getState().secret = true
```

Comment:
The 4th parameter of `toOutputWithContext` is `forceSecretVal *bool`, and it forces the value to be a secret on resolution:

https://github.com/pulumi/pulumi/blob/0b13e67b0a89f3ee03ca3e243e70a93386653a8e/sdk/go/pulumi/types.go#L895-L914

However, if this feels race-y still, line 905-907 could be moved above the `go func()` here and directly set output.getState().secret. 

That said, I think `secret`, `known`, and `deps` should all be considered illegal to read from until awaited. Any code that reads those fields before the await will have incomplete information in the case of nested outputs:

```go
// unwrap an Output[Output[T]]
someOutput.ApplyT(func (x AnyOutput) {
  return x
})
```

Or a concrete example, summing two int outputs without using .All():

```go
xOutput.ApplyT(func (x int) IntOutput {
  return yOutput.ApplyT(func (y int) IntOutput {
    return x + y
  }).(IntOutput)
}).(IntOutput)
```

This latter output has the combined known-ness, secret-ness, and dependencies of both xOutput and yOutput, but that can't be known until the outer output is awaited (which will await the inner and accumulate.)

Reply:
My concern with this change is less about the raciness of the boolean toggle, but more on relying on `toOutputWithContext` to guarantee the value is secret. I'd hate for someone to change the behavior of `toOutputWithContext` and inadvertently render `ToSecretWithContext` useless. 

I have to imagine catching a bug like that is the job of unit testing and we shouldn't worry about that here 😅

Reply:
But I also do think you're on to something with moving line 905-907 above the` go func()`. I agree that it should be illegal to read from a value before its awaited, but I also think users will do all kinds of screwy things with `Apply`, and the SecOps team would probably prefer we keep secrets secret even when their developers cause trouble. My gut says its less of a foot-gun to set the value outside of the Goroutine.

Reply:
I checked and we do have tests for this that fail if the line is removed - secrets that fail to warn that they're secret.

Even so, I agree with "initializing" the output with secret=true before the goroutine call, because I cannot prove there aren't other places in our code where we've illegally read from the secret field before awaiting.

Reply:
That's a sufficient resolution for me then! Just re-request me when you make that change and I'll give you the hammer stamp! 🔨  Thanks Friel, this is super nice, definitely helps contribute towards our company's Q3 mission. :D

Reply:
I'll also add a comment on toOutputWithContext that the forceSecret argument is load bearing and ensures that the output is marked secret before any other code could possibly await or read that field.